src/core/Akka.Remote/EndpointManager.cs

```diff
@@ -492,196 +543,248 @@ protected void Accepting(object message)
+                        // B: we know the UID, but it does not match with the UID to be quarantined
                     }
 
-                    //Stop inbound read-only association
-                    var read = endpoints.ReadOnlyEndpointFor(quarantine.RemoteAddress);
-                    if (read != null)
+                    // Stop inbound read-only associations
+                    var readPolicy = Tuple.Create(_endpoints.ReadOnlyEndpointFor(quarantine.RemoteAddress), quarantine.Uid);
+                if (readPolicy.Item1?.Item1 != null && quarantine.Uid == null)
+                    Context.Stop(readPolicy.Item1.Item1);
+                else if (readPolicy.Item1?.Item1 != null && quarantine.Uid != null && readPolicy.Item1?.Item2 == quarantine.Uid) { Context.Stop(readPolicy.Item1.Item1); }
+                else { } // nothing to stop
+
+                    Func<AkkaProtocolHandle, bool> matchesQuarantine = handle => handle.RemoteAddress.Equals(quarantine.RemoteAddress) &&
+                                                                             quarantine.Uid == handle.HandshakeInfo.Uid;
+
+                    // Stop all matching pending read handoffs
+                    _pendingReadHandoffs = _pendingReadHandoffs.Where(x =>
+                {
+                    var drop = matchesQuarantine(x.Value);
+                        // Side-effecting here
+                        if (drop)
                     {
-                        Context.Stop((IInternalActorRef)read);
+                        x.Value.Disassociate();
+                        Context.Stop(x.Key);
                     }
+                    return !drop;
+                }).ToDictionary(key => key.Key, value => value.Value);
 
-                    if (quarantine.Uid.HasValue)
+                    // Stop all matching stashed connections
+                    _stashedInbound = _stashedInbound.Select(x =>
+                {
+                    var associations = x.Value.Where(assoc =>
                     {
-                        endpoints.MarkAsQuarantined(quarantine.RemoteAddress, quarantine.Uid.Value, Deadline.Now + settings.QuarantineDuration);
-                        eventPublisher.NotifyListeners(new QuarantinedEvent(quarantine.RemoteAddress, quarantine.Uid.Value));
-                    }
-                })
-                .With<Send>(send =>
+                        var handle = assoc.Association.AsInstanceOf<AkkaProtocolHandle>();
+                        var drop = matchesQuarantine(handle);
+                        if (drop)
+                            handle.Disassociate();
+                        return !drop;
+                    }).ToList();
+                    return new KeyValuePair<IActorRef, List<InboundAssociation>>(x.Key, associations);
+                }).ToDictionary(k => k.Key, v => v.Value);
+
+                if (quarantine.Uid.HasValue)
```

Comment:
Finally, if the `Quarantine` message has a UID, mark it as `Quarantined` in the `EndpointRegistry` and publish an event letting subscribers know about the affected endpoint.